core-java/src/main/java/eu/neverblink/jelly/core/internal/ProtoEncoderImpl.java

```diff
@@ -0,0 +1,133 @@
+ * Take care to ensure the correctness of the transmitted data, or use the specialized wrappers from the stream package.
+ */
+public class ProtoEncoderImpl<TNode> extends ProtoEncoder<TNode> {
+
+    private boolean hasEmittedOptions = false;
+    private final Collection<RdfStreamRow> rowBuffer;
+
+    /**
+     * Constructor for the ProtoEncoderImpl class.
+     * <p>
+     *
+     * @param converter converter for the encoder
+     * @param params    parameters object for the encoder
+     */
+    public ProtoEncoderImpl(ProtoEncoderConverter<TNode> converter, ProtoEncoder.Params params) {
+        super(converter, params);
+        this.rowBuffer = appendableRowBuffer;
+    }
+
+    @Override
+    public void handleTriple(TNode subject, TNode predicate, TNode object) {
+        emitOptions();
+        final var triple = tripleToProto(subject, predicate, object);
+        final var mainRow = RdfStreamRow.newBuilder().setTriple(triple.toProto()).build();
+        rowBuffer.add(mainRow);
+    }
+
+    @Override
+    public void handleQuad(TNode subject, TNode predicate, TNode object, TNode graph) {
+        emitOptions();
+        final var quad = quadToProto(subject, predicate, object, graph);
+        final var mainRow = RdfStreamRow.newBuilder().setQuad(quad.toProto()).build();
+        rowBuffer.add(mainRow);
+    }
+
+    @Override
+    public void handleGraphStart(TNode graph) {
+        emitOptions();
+        final var graphNode = converter.graphNodeToProto(nodeEncoder, graph);
+        final var graphStart = new RdfTerm.GraphStart(graphNode);
+        final var graphRow = RdfStreamRow.newBuilder().setGraphStart(graphStart.toProto()).build();
+        rowBuffer.add(graphRow);
+    }
+
+    @Override
+    public void handleGraphEnd() {
+        if (!hasEmittedOptions) {
+            throw new RdfProtoSerializationError("Cannot end a delimited graph before starting one");
+        }
+
+        final var graphEnd = new RdfTerm.GraphEnd();
+        final var graphRow = RdfStreamRow.newBuilder().setGraphEnd(graphEnd.toProto()).build();
+        rowBuffer.add(graphRow);
+    }
+
+    @Override
+    public void handleNamespace(String prefix, TNode namespace) {
+        if (!enableNamespaceDeclarations) {
+            throw new RdfProtoSerializationError("Namespace declarations are not enabled in this stream");
+        }
+
+        emitOptions();
+
+        final var namespaceTerm = converter.nodeToProto(nodeEncoder, namespace);
+        if (!(namespaceTerm instanceof RdfTerm.Iri iriTerm)) {
+            throw new RdfProtoSerializationError("Namespace must be an IRI");
+        }
+
+        final var mainRow = RdfStreamRow.newBuilder()
+            .setNamespace(RdfNamespaceDeclaration.newBuilder().setName(prefix).setValue(iriTerm.toProto()).build())
+            .build();
+
+        rowBuffer.add(mainRow);
+    }
+
+    @Override
+    public void handleNamespace(String prefix, String namespace) {
```

Comment:
this is duplicated from the above :D no need to have two, keep the one defined in RdfHandler